src/vision/status_quo/barbara_bridges_sync_and_async.md

```diff
@@ -0,0 +1,245 @@
+            .collect();
+}
+```
+
+
+but now she gets a type error again:
+
+```
+error[E0277]: a value of type `Vec<Data>` cannot be built from an iterator over elements of type `tokio::task::JoinHandle<impl futures::Future>`
+  --> src/main.rs:22:14
+   |
+22 |             .collect();
+   |              ^^^^^^^ value of type `Vec<Data>` cannot be built from `std::iter::Iterator<Item=tokio::task::JoinHandle<impl futures::Future>>`
+   |
+   = help: the trait `FromIterator<tokio::task::JoinHandle<impl futures::Future>>` is not implemented for `Vec<Data>`
+```
+
+Of course! `spawn_blocking`, like `map`, just takes a regular closure, not an async closure. She's getting a bit frustrated now. "Well," she thinks, "I can use `spawn` to get into an async context!" So she adds a call to `spawn` inside the `spawn_blocking` closure:
+
+```rust
+fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> =
+        urls
+            .iter()
+            .map(|url| tokio::task::spawn_blocking(move || {
+                tokio::task::spawn(async move {
+                    do_web_request(url).await
+                })
+            }))
+            .collect();
+}
+```
+
+But this isn't really helping, as `spawn` still yields a future. She's getting the same errors.
+
+### Async all the way
+
+She remembers now that this whole drama started because she was converting her `main` function to be `async`. Maybe she doesn't have to bridge between sync and async? She starts digging around in the docs and finds `futures::join_all`. Using that, she can change `aggregate` to be an async function too:
+
+```rust
+async fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> = futures::join_all(
+        urls
+            .iter()
+            .map(|url| do_web_request(url))
+    ).await;
+}
+```
+
+Things are working again now, so she is happy.
+
+### Filtering
+
+Later on, she would like to apply a filter to the aggregation operation. She realizes that if she wants to use the fetched data when doing the filtering, she has to filter the vector after the join has completed; `join_all` doesn't have a way to put a filter into the iterator chain like she wants. She is annoyed, but performance isn't critical, so it's ok.
+
+### And the cycle begins again
+
+Later on, she wants to call `aggregate` from another binary. This one doesn't have an `async main`. This context is deep inside of an iterator chain and was previously entirely synchronous. She realizes it would be a lot of work to change all the intervening stack frames to be `async fn`, rewrite the iterators into streams, etc. She decides to just call `block_on` again, even though it make her nervous.
+
+## 🤔 Frequently Asked Questions
+
+### What are the morals of the story?
+
+* Some projects don't care about max performance and just want things to work.
+    * They would probably be happy with sync but as the most popular libraries for web requests, databases, etc, offer async interfaces, they may still be using async code.
```

Comment:
`reqwest` offers a sync API. Does Barbara not know about this? Or does she know but wishes to use async because she plans on moving her code to be async?

Reply:
We actually discuss this below -- even if she did use the sync API, she'd eventually have the same issues. However, it's a good question why she didn't to begin with. I think in the "real story" she's already using a mix of sync and async in various places and the API itself wasn't reqwest but rather something related to databases which did not necessarily have a sync variant. Perhaps I should just change from reqwest to that.

Reply:
FWIW my recollection - not sure how much this is a mix of different experiences - but the reqwest sync API at least from my POV wasn't used because (a) you need to enable the blocking feature, it's not on by default - so additional hurdle and (b) I had vague poor experience - I think documented elsewhere in these stories - with panics when mixing blocking (sync) with async APIs.

Reply:
In my experience, not all the case you can control whether the code is sync or async. Some libraries will build a sync-async later and only accept a sync layer from you. I was doing tokio-rustls code, I have a runtime at the main code and I only get to build on top of a sync code. So it is `async <-> sync <-> async` and the only part I can change is the last part (to avoid trouble), one reason could be that there is no `async fn` for `impl` in stable rust yet. Sample of similar code I worked with (I found the docs for low-level async code not as well written or less examples as compared to the sync code in other parts of rust):

https://github.com/FlorianUekermann/rustls-acme/blob/a3f31ca9a79b6b031dd8ea3d68777ee838f98675/src/simple.rs#L28
https://github.com/FlorianUekermann/rustls-acme/blob/a3f31ca9a79b6b031dd8ea3d68777ee838f98675/src/resolver.rs#L240

But usually, if I only do one request I would probably avoid using async and just use reqwest sync API because it's way easier without anything to look out for, enabling the blocking feature is definitely easier than dealing with all kinds of async foot guns.

> I had vague poor experience - I think documented elsewhere in these stories - with panics when mixing blocking (sync) with async APIs.

Mixing sync with async is okay, same for async with sync. But if you want async within sync within async (hamburger) then the trouble starts.

Reply:
I added a FAQ to attempt to answer this.